single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,749 @@
+            self.mongos_config_manager,
+            self.backup_manager,
+            self.logrotate_config_manager,
+            self.mongodb_exporter_config_manager,
+        )
+
+    # BEGIN: Handlers.
+
+    @override
+    def on_install(self) -> None:
+        """Handler on install."""
+        if not self.workload.snap_present:
+            return
+        if not self.workload.container_can_connect:
+            raise ContainerNotReadyError
+        self.charm.unit.set_workload_version(self.workload.get_version())
+
+        # Truncate the file.
+        self.workload.write(self.workload.paths.config_file, "")
+
+        for config_manager in self.config_managers:
+            config_manager.set_environment()
+
+        self.logrotate_config_manager.connect()
+
+    @override
+    def on_start(self) -> None:
+        """Handler on start."""
+        if not self.workload.container_can_connect:
+            logger.debug("mongod container is not ready yet.")
+            raise ContainerNotReadyError
+
+        if any(not storage for storage in self.model.storages.values()):
+            logger.debug("Storages not attached yet.")
+            raise ContainerNotReadyError
+
+        self.instantiate_keyfile()
+        self.tls_manager.push_tls_files_to_workload()
+        self.handle_licenses()
+        self.set_permissions()
+
+        try:
+            logger.info("Starting MongoDB.")
+            self.charm.status_manager.to_maintenance("starting MongoDB")
+            self.start_charm_services()
+            self.charm.status_manager.to_active(None)
+        except WorkloadServiceError as e:
+            logger.error(f"An exception occurred when starting mongod agent, error: {e}.")
+            self.charm.status_manager.to_blocked("couldn't start MongoDB")
+            return
+
+        # Open ports:
+        try:
+            self.open_ports()
+        except WorkloadExecError:
+            self.charm.status_manager.to_blocked("failed to open TCP port for MongoDB")
+            raise
+
+        if self.substrate == Substrates.K8S:
+            if not self.workload.exists(self.workload.paths.socket_path):
+                logger.debug("The mongod socket is not ready yet.")
+                raise WorkloadNotReadyError
+
+        if not self.mongo_manager.mongod_ready():
+            self.charm.status_manager.to_waiting("waiting for MongoDB to start")
+            raise WorkloadNotReadyError
+
+        self.charm.status_manager.to_active(None)
+
+        try:
+            self.mongodb_exporter_config_manager.connect()
+        except WorkloadServiceError:
+            self.charm.status_manager.to_blocked("couldn't start mongodb exporter")
+            return
+
+        try:
+            self.backup_manager.connect()
+        except WorkloadServiceError:
+            self.charm.status_manager.to_blocked("couldn't start pbm-agent")
+            return
+
+        self._initialise_replica_set()
+        self.charm.status_manager.to_active(None)
```

Comment:
In general - I'm not a fan of setting to `active`. This can be the object of another PR when Mia finds a suitable solution for statuses.

Reply:
As I said in the PR description, this is very basic an currently only exists for progress purposes and testing.
I haven't worked at all on statuses apart from having something that allows me to set it easily.
This will be the object of a future PR during next pulse.